Subsidy API implementation and Integration test

DedrickEnc opened the pull request:
In this PR, I implemented the subsidy API and tested all routes
The subsidy API has methods :
POST
GET
PUT
DELETE

The codes.js was updated to add more error cases, the routes.js file was updated too.

jniles commented on server/config/routes.js:
```diff
@@ -135,7 +137,14 @@ exports.configure = function (app) {
+
+ //API for subsidy routes crud
+  app.get('/subsidies', subsidy.list);
+  app.get('/subsidies/:id', subsidy.getSubsidy);
```
As discussed, this should be `sudsidy.detail`.

jniles commented on server/controllers/finance/subsidy.js:
```diff
@@ -0,0 +1,114 @@
+
+  if(isEmptyObject(obj)) { throw codes.ERR_EMPTY_BODY;}
+  if(!obj.value) { throw codes.ERR_PARAMETERS_REQUIRED;}
+  if(obj.value <= 0) { throw codes.ERR_BAD_VALUE;}
```
It would help to have a more specific error message here. Why is this a bad value?

DedrickEnc replied:
The test is clear, it is less than zero, so that value is not correct.

jniles replied:
The test is very clear.  The problem is that an API consumer shouldn't have to have the application source code to know why an error occurred.  A helpful error message would save a lot of time and pain for that developer.

jniles commented on server/controllers/finance/subsidy.js:
```diff
@@ -0,0 +1,114 @@
+function checkData (obj, codes){
+
+  if(isEmptyObject(obj)) { throw codes.ERR_EMPTY_BODY;}
+  if(!obj.value) { throw codes.ERR_PARAMETERS_REQUIRED;}
```
Hmm..  The ERR_PARAMETERS_REQUIRED was originally designed for HTTP parameters - query strings.  I see now that the error message is fairly generic.  It might be better to split these errors into two: 1) missing query parameters (used in @mbayopanda's patient search API) and 2) missing req.body properties for UPDATEs/CREATEs.

DedrickEnc replied:
As many PR is using this ERR_PARAMETERS_REQUIRED property, it can be corrected as separated pull request.

jniles replied:
Sounds good.

jniles commented on server/test/api/subsidy.js:
```diff
@@ -0,0 +1,112 @@
+helpers.configure(chai);
+
+var url = 'https://localhost:8080';
+var user = { username : 'superuser', password : 'superuser', project: 1};
```
These two variables are unused.

DedrickEnc replied:
you are right

jniles commented on server/test/api/subsidy.js:
```diff
@@ -0,0 +1,112 @@
+
+  beforeEach(helpers.login(agent));
+
+  it('METHOD : POST, PATH : /subsidies, It adds a subsidy', function () {
```
Good test!

jniles commented on server/test/api/subsidy.js:
```diff
@@ -0,0 +1,112 @@
+      })
+      .then(function (res){ 
+        expect(res).to.have.status(200);
+        expect(res.body).to.have.all.keys('id', 'account_id', 'label', 'description', 'value', 'created_at', 'updated_at');
```
Since you reuse these keys over and over, you might want to move them into their own variable at the top of your `describe()` block.

DedrickEnc replied:
Be more clear, please!

jniles replied:
``` js
// just after the function definition for `describe()`
var keys = ['id', 'account_id', 'label', 'description', 'value', 'created_at', 'updated_at'];

// later in a  test
expect(res.body).to.have.all.keys(keys);
```

jniles commented on server/test/api/subsidy.js:
```diff
@@ -0,0 +1,112 @@
+  it('METHOD : POST, PATH : /subsidies, It refuses to add a wrong subsidy', function () {
+    return agent.post('/subsidies')
+      .send(wrongSubsidy)
+      .then(function (res) {
```
Another thing to test for: what should the error code (sent in the JSON object) be?

jniles commented on server/test/api/subsidy.js:
```diff
@@ -0,0 +1,112 @@
+        .catch(helpers.handler);
+    });
+  
+  it('METHOD : GET, PATH : /subsidies/:id, It returns one subsidy', function () {
```
This test is a duplicate of the second half of 'It adds a new subsidy' (Line 33).  You don't need it.

DedrickEnc replied:
I understand, but no matter, it is more informative, here the `/subsidies/:id` is tested directly.
Each route must be tested, directly and separately, `subsidies/:id` is one of them.

jniles replied:
Duplicate tests aren't a terrible problem.  One caution though: if the POST method breaks, you may spend lots of time trying to debug the GET request rather than the POST request, since it will look like the GET request is failing as well.

jniles commented on server/test/api/subsidy.js:
```diff
@@ -0,0 +1,112 @@
+        expect(res).to.have.status(200);
+        expect(res).to.be.json;
+        expect(res.body.id).to.equal(newSubsidy.id);
+        expect(res.body.value).to.equal(updateInfo.value);
```
It would be useful to check that other properties didn't change.

DedrickEnc replied:
Why?

jniles replied:
Why not?

jniles commented on server/test/api/subsidy.js:
```diff
@@ -0,0 +1,112 @@
+        // re-query the database
+        return agent.get('/subsidies/' + newSubsidy.id);
+      })
+      .then(function (res) {
```
If you want to be _super_ sure that this worked correctly, you can also query `/subsidies` and check that the length is back to being the original length expected.

jniles commented on server/test/api/subsidy.js:
```diff
@@ -0,0 +1,112 @@
+  };
+
+  beforeEach(helpers.login(agent));
+
```
One more test to add: GET /subsidies when there are no subsidies in the database should return a 200 status and empty array.

DedrickEnc replied:
For common issue like this, a separate pull request will be good, because almost all our test is dealing like this, so a separate pull request will walk through all those file and correct them if necessary.

jniles replied:
I'll fix the other tests in the `test-handlers` PR if you'll do this one, that would be very helpful.

jniles commented on server/test/api/subsidy.js:
```diff
@@ -0,0 +1,112 @@
+/*global describe, it, beforeEach, process*/
```
You don't need "process" in this comment.

DedrickEnc committed:
Subsidy API implementation and Integration test

In this PR, I implemented the subsidy API and tested all routes
The subsidy API has methods :
POST
GET
PUT
DELETE

The codes.js was updated to add more error cases, the routes.js file was updated too.

DedrickEnc committed:
Subsidy API implementation and Integration test

In this PR, I implemented the subsidy API and tested all routes
The subsidy API has methods :
POST
GET
PUT
DELETE

The codes.js was updated to add more error cases, the routes.js file was updated too.

DedrickEnc committed:
amending the PR

DedrickEnc committed:
amending continues
